src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -348,7 +348,9 @@ public Next receive(Object o) {
             for (BodyExecutionCallback c : callbacks) {
                 c.onFailure(sc, t);
             }
-
+            synchronized (CpsBodyExecution.this) {
+                thread.popContextVariables();
```

Comment:
`thread` might be null here (see `!isLaunched()` clause above). I will fix that in #279.